lib/Sema/TypeCheckType.cpp

```diff
@@ -1325,6 +1260,26 @@ resolveTopLevelIdentTypeComponent(TypeResolution resolution,
     if (options.contains(TypeResolutionFlags::SilenceErrors))
       return ErrorType::get(ctx);
 
+    if (comp->getIdentifier() == ctx.Id_Self &&
+        !isa<GenericIdentTypeRepr>(comp)) {
+      DeclContext *nominalDC = nullptr;
+      NominalTypeDecl *nominal = nullptr;
+      auto dc = resolution.getDeclContext();
+      if ((nominalDC = dc->getInnermostTypeContext()) &&
+          (nominal = nominalDC->getSelfNominalTypeDecl())) {
+        // Attempt to refer to 'Self' within a non-protocol nominal
+        // type. Fix this by replacing 'Self' with the nominal type name.
+        assert(!isa<ProtocolDecl>(nominal) && "Cannot be a protocol");
+
+        return nominal->getDeclaredInterfaceType();
```

Comment:
You need to do `resolution.mapTypeIntoContext(nominal->getDeclaredInterfaceType())` here. This should fix your issue with casting to `Self`.

Reply:
Also, this should wrap the type in a DynamicSelfType, just like you do in expression context. Otherwise, the following will have different behavior:

```
let a = Self.init()
let b: Self = Self.init()
```

Reply:
Brilliant! The mapTypeIntoContext fixed the problems with casting. I don't see what you mean about DynamicSelfType. I tried something but it gave an assertion failure (see comment)